bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java

```diff
@@ -95,72 +99,112 @@ public Locale locale() {
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided sink
-     */
-    public DialogContext withSink(AudioSink sink) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided stt
-     */
-    public DialogContext withSTT(STTService stt) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided tts
-     */
-    public DialogContext withTTS(TTSService tts) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided voice
-     */
-    public DialogContext withVoice(Voice voice) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided list of interpreters
-     */
-    public DialogContext withHLIs(List<HumanLanguageInterpreter> hlis) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided locale
-     */
-    public DialogContext withLocale(Locale locale) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided listening item
-     */
-    public DialogContext withListeningItem(String listeningItem) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided ks
-     */
-    public DialogContext withKS(@Nullable KSService ks) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided keyword
-     */
-    public DialogContext withKeyword(@Nullable String keyword) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
+    public static class Builder {
+        private final Logger logger = LoggerFactory.getLogger(Builder.class);
+        // services
+        private @Nullable AudioSource source = null;
+        private @Nullable AudioSink sink = null;
+        private @Nullable KSService ks = null;
+        private @Nullable STTService stt = null;
+        private @Nullable TTSService tts = null;
+        private @Nullable Voice voice = null;
+        private List<HumanLanguageInterpreter> hlis = List.of();
+        // options
+        private @Nullable String listeningItem = null;
```

Comment:
I think all these null assignments are rudundant?

Reply:
Not sure about it, I will checked it.

Reply:
Yes there are, nice to know, I think I did this in a bunch of places, thank you!

Reply:
I once [read](http://www.javapractices.com/topic/TopicAction.do?Id=14) default value assignment causes more byte code to be generated and it also adds more cognitive load because it's more code that has to be read. Also you'll encounter unassigned values all over the place, so you have to know the default values anyhow.

Reply:
Oh, really nice article, I was thinking about it and I guess I tend to initialize the values because I program a lot on typescript and it blame me to initialize everything (due to the dynamic types in javascript) but I don't think I forget this now, thank you!!